From 617864e9703de4f8bfe8766106e17a7ed79bee8f Mon Sep 17 00:00:00 2001
From: Joel Hockey <joelhockey@chromium.org>
Date: Thu, 31 Jan 2019 19:15:25 -0800
Subject: [PATCH 5/7] shared.Unpack: Add support for a ProgressTracker during
 unpack

This is part 3 of a series of patches to add better progress
tracking support for export and import.

Signed-off-by: Joel Hockey <joelhockey@chromium.org>
---
 lxd/images.go            |  7 ++++---
 lxd/storage_btrfs.go     |  2 +-
 lxd/storage_ceph.go      |  2 +-
 lxd/storage_dir.go       |  2 +-
 lxd/storage_lvm.go       |  2 +-
 lxd/storage_lvm_utils.go |  2 +-
 lxd/storage_zfs.go       |  2 +-
 shared/archive_linux.go  | 42 ++++++++++++++++++++++++++++++----------
 8 files changed, 42 insertions(+), 19 deletions(-)

diff --git a/lxd/images.go b/lxd/images.go
index 8fa04e59..7a058976 100644
--- a/lxd/images.go
+++ b/lxd/images.go
@@ -32,6 +32,7 @@ import (
 	"github.com/lxc/lxd/lxd/util"
 	"github.com/lxc/lxd/shared"
 	"github.com/lxc/lxd/shared/api"
+	"github.com/lxc/lxd/shared/ioprogress"
 	"github.com/lxc/lxd/shared/logger"
 	"github.com/lxc/lxd/shared/logging"
 	"github.com/lxc/lxd/shared/osarch"
@@ -97,14 +98,14 @@ var aliasCmd = Command{
    end for whichever finishes last. */
 var imagePublishLock sync.Mutex
 
-func unpackImage(imagefname string, destpath string, sType storageType, runningInUserns bool) error {
+func unpackImage(imagefname string, destpath string, sType storageType, runningInUserns bool, tracker *ioprogress.ProgressTracker) error {
 	blockBackend := false
 
 	if sType == storageTypeLvm || sType == storageTypeCeph {
 		blockBackend = true
 	}
 
-	err := shared.Unpack(imagefname, destpath, blockBackend, runningInUserns)
+	err := shared.Unpack(imagefname, destpath, blockBackend, runningInUserns, tracker)
 	if err != nil {
 		return err
 	}
@@ -116,7 +117,7 @@ func unpackImage(imagefname string, destpath string, sType storageType, runningI
 			return fmt.Errorf("Error creating rootfs directory")
 		}
 
-		err = shared.Unpack(imagefname+".rootfs", rootfsPath, blockBackend, runningInUserns)
+		err = shared.Unpack(imagefname+".rootfs", rootfsPath, blockBackend, runningInUserns, tracker)
 		if err != nil {
 			return err
 		}
diff --git a/lxd/storage_btrfs.go b/lxd/storage_btrfs.go
index 42350adc..41d72bfb 100644
--- a/lxd/storage_btrfs.go
+++ b/lxd/storage_btrfs.go
@@ -1468,7 +1468,7 @@ func (s *storageBtrfs) ImageCreate(fingerprint string) error {
 
 	// Unpack the image in imageMntPoint.
 	imagePath := shared.VarPath("images", fingerprint)
-	err = unpackImage(imagePath, tmpImageSubvolumeName, storageTypeBtrfs, s.s.OS.RunningInUserNS)
+	err = unpackImage(imagePath, tmpImageSubvolumeName, storageTypeBtrfs, s.s.OS.RunningInUserNS, nil)
 	if err != nil {
 		return err
 	}
diff --git a/lxd/storage_ceph.go b/lxd/storage_ceph.go
index 90f197b5..3b793238 100644
--- a/lxd/storage_ceph.go
+++ b/lxd/storage_ceph.go
@@ -1891,7 +1891,7 @@ func (s *storageCeph) ImageCreate(fingerprint string) error {
 
 		// rsync contents into image
 		imagePath := shared.VarPath("images", fingerprint)
-		err = unpackImage(imagePath, imageMntPoint, storageTypeCeph, s.s.OS.RunningInUserNS)
+		err = unpackImage(imagePath, imageMntPoint, storageTypeCeph, s.s.OS.RunningInUserNS, nil)
 		if err != nil {
 			logger.Errorf(`Failed to unpack image for RBD storage volume for image "%s" on storage pool "%s": %s`, fingerprint, s.pool.Name, err)
 
diff --git a/lxd/storage_dir.go b/lxd/storage_dir.go
index e6739edd..cbf09f50 100644
--- a/lxd/storage_dir.go
+++ b/lxd/storage_dir.go
@@ -508,7 +508,7 @@ func (s *storageDir) ContainerCreateFromImage(container container, imageFingerpr
 	}()
 
 	imagePath := shared.VarPath("images", imageFingerprint)
-	err = unpackImage(imagePath, containerMntPoint, storageTypeDir, s.s.OS.RunningInUserNS)
+	err = unpackImage(imagePath, containerMntPoint, storageTypeDir, s.s.OS.RunningInUserNS, nil)
 	if err != nil {
 		return err
 	}
diff --git a/lxd/storage_lvm.go b/lxd/storage_lvm.go
index 9e62b576..971387e7 100644
--- a/lxd/storage_lvm.go
+++ b/lxd/storage_lvm.go
@@ -1618,7 +1618,7 @@ func (s *storageLvm) ImageCreate(fingerprint string) error {
 		}
 
 		imagePath := shared.VarPath("images", fingerprint)
-		err = unpackImage(imagePath, imageMntPoint, storageTypeLvm, s.s.OS.RunningInUserNS)
+		err = unpackImage(imagePath, imageMntPoint, storageTypeLvm, s.s.OS.RunningInUserNS, nil)
 		if err != nil {
 			return err
 		}
diff --git a/lxd/storage_lvm_utils.go b/lxd/storage_lvm_utils.go
index 565ac58d..f6b24c6d 100644
--- a/lxd/storage_lvm_utils.go
+++ b/lxd/storage_lvm_utils.go
@@ -458,7 +458,7 @@ func (s *storageLvm) containerCreateFromImageLv(c container, fp string) error {
 
 	imagePath := shared.VarPath("images", fp)
 	containerMntPoint := getContainerMountPoint(s.pool.Name, containerName)
-	err = unpackImage(imagePath, containerMntPoint, storageTypeLvm, s.s.OS.RunningInUserNS)
+	err = unpackImage(imagePath, containerMntPoint, storageTypeLvm, s.s.OS.RunningInUserNS, nil)
 	if err != nil {
 		logger.Errorf(`Failed to unpack image "%s" into non-thinpool LVM storage volume "%s" for container "%s" on storage pool "%s": %s`, imagePath, containerMntPoint, containerName, s.pool.Name, err)
 		return err
diff --git a/lxd/storage_zfs.go b/lxd/storage_zfs.go
index d97088e5..1a6e18ee 100644
--- a/lxd/storage_zfs.go
+++ b/lxd/storage_zfs.go
@@ -1787,7 +1787,7 @@ func (s *storageZfs) ImageCreate(fingerprint string) error {
 	}
 
 	// Unpack the image into the temporary mountpoint.
-	err = unpackImage(imagePath, tmpImageDir, storageTypeZfs, s.s.OS.RunningInUserNS)
+	err = unpackImage(imagePath, tmpImageDir, storageTypeZfs, s.s.OS.RunningInUserNS, nil)
 	if err != nil {
 		return err
 	}
diff --git a/shared/archive_linux.go b/shared/archive_linux.go
index bc3ebe00..8c6a4414 100644
--- a/shared/archive_linux.go
+++ b/shared/archive_linux.go
@@ -3,10 +3,12 @@ package shared
 import (
 	"bytes"
 	"fmt"
+	"io"
 	"os"
 	"strings"
 	"syscall"
 
+	"github.com/lxc/lxd/shared/ioprogress"
 	"github.com/lxc/lxd/shared/logger"
 )
 
@@ -50,7 +52,7 @@ func DetectCompression(fname string) ([]string, string, error) {
 
 }
 
-func Unpack(file string, path string, blockBackend bool, runningInUserns bool) error {
+func Unpack(file string, path string, blockBackend bool, runningInUserns bool, tracker *ioprogress.ProgressTracker) error {
 	extractArgs, extension, err := DetectCompression(file)
 	if err != nil {
 		return err
@@ -58,6 +60,7 @@ func Unpack(file string, path string, blockBackend bool, runningInUserns bool) e
 
 	command := ""
 	args := []string{}
+	var reader io.Reader
 	if strings.HasPrefix(extension, ".tar") {
 		command = "tar"
 		if runningInUserns {
@@ -69,8 +72,32 @@ func Unpack(file string, path string, blockBackend bool, runningInUserns bool) e
 		}
 		args = append(args, "-C", path, "--numeric-owner", "--xattrs-include=*")
 		args = append(args, extractArgs...)
-		args = append(args, file)
+		args = append(args, "-")
+
+		f, err := os.Open(file)
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+
+		reader = f
+
+		// Attach the ProgressTracker if supplied.
+		if tracker != nil {
+			fsinfo, err := f.Stat()
+			if err != nil {
+				return err
+			}
+
+			tracker.Length = fsinfo.Size()
+			reader = &ioprogress.ProgressReader{
+				ReadCloser: f,
+				Tracker:    tracker,
+			}
+		}
 	} else if strings.HasPrefix(extension, ".squashfs") {
+		// unsquashfs does not support reading from stdin,
+		// so ProgressTracker is not possible.
 		command = "unsquashfs"
 		args = append(args, "-f", "-d", path, "-n")
 
@@ -87,7 +114,7 @@ func Unpack(file string, path string, blockBackend bool, runningInUserns bool) e
 		return fmt.Errorf("Unsupported image format: %s", extension)
 	}
 
-	output, err := RunCommand(command, args...)
+	err = RunCommandWithFds(reader, nil, command, args...)
 	if err != nil {
 		// Check if we ran out of space
 		fs := syscall.Statfs_t{}
@@ -106,14 +133,9 @@ func Unpack(file string, path string, blockBackend bool, runningInUserns bool) e
 			}
 		}
 
-		co := output
 		logger.Debugf("Unpacking failed")
-		logger.Debugf(co)
-
-		// Truncate the output to a single line for inclusion in the error
-		// message.  The first line isn't guaranteed to pinpoint the issue,
-		// but it's better than nothing and better than a multi-line message.
-		return fmt.Errorf("Unpack failed, %s.  %s", err, strings.SplitN(co, "\n", 2)[0])
+		logger.Debugf(err.Error())
+		return fmt.Errorf("Unpack failed, %s.", err)
 	}
 
 	return nil
-- 
2.20.1.611.gfbb209baf1-goog

